sdl-core/src/main/scala/io/smartdatalake/workflow/StateMigratorDef.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Smart Data Lake - Build your data lake the smart way.
+ *
+ * Copyright © 2019-2023 ELCA Informatique SA (<https://www.elca.ch>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package io.smartdatalake.workflow
+
+import io.smartdatalake.util.misc.SmartDataLakeLogger
+import org.json4s.JsonAST.JInt
+import org.json4s.{JArray, JValue}
+
+/**
+ * Definition of how to migrate from one state version to another.
+ * Note that this should always migrate an existing version to the next version, without skipping a version.
+ * There might be other state migrators applied afterwards to migrate to the current version.
+ */
+private[smartdatalake] trait StateMigratorDef {
+  def versionFrom: Int
+  def versionTo: Int
+  def migrate(json: JValue): JValue
+  override def toString: String = s"$versionFrom -> $versionTo"
+}
+
+/**
+ * Migrate state from format version 3 to 4:
+ * - the structure of `actionState.results` was cleaned up.
+ *   instead of having `results.subFeed`, the attributes of `subFeed` have been moved directly into `result`.
+ * - Additionally `actionState.inputIds[].id` was cleaned up as `inputIds[]`, and the the same applies for `actionState.outputIds`.
+ */
+class StateMigratorDef3To4 extends StateMigratorDef with SmartDataLakeLogger {
+  override val versionFrom = 3
+  override val versionTo = 4
+  override def migrate(json: JValue): JValue = {
+    assert(json \ "runStateFormatVersion" match { case JInt(version) => version <= versionFrom}, s"Version should be equals or less than $versionFrom")
```

Comment:
Shouldn't we assume that version = versionFrom ? The upgrade procedure is probably strictly working for one version only.

Reply:
Then we could not migrate from version 1/2 to version 4. As there are no changes needed to migrate from version 1to2 or 2to3, no StateMigratorDef exists for these steps.

Reply:
OK, makes sense. Thanks!